### PR TITLE
Fixed gradient checkpoint bug for TimeSeriesTransformer

### LINKED ISSUE #21737: [`Generate`] Fix `gradient_checkpointing` and `use_cache` bug for generate-compatible models
## Feature request

When using a model that uses `gradient_checkpointing` and if a user wants to call `generate` with `use_cache`, it leads some models to bugs, such as the one described in https://github.com/huggingface/transformers/pull/21733

The fix should be to slightly refactor some models following the same procedure as in the aforementioned PR

### How to participate

1. If it is your first time here, have a quick look at our [contribution guidelines](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md) 🤗 
2. Pick a model from the list below. Check in the comments here if it hasn't been claimed yet.
3. Claim your models in the comments (e.g. "I want to work on GPT2")
4. Replicate the changes of [this PR](https://github.com/huggingface/transformers/pull/21733) to your model of choice. In other words, move the `if` block to the line above the `... if use_cache else None`, in the same `.forward()` function. Please note that some models may have more than one instance of this block!
5. Make sure you've run our automated code formatting tool (i.e. run `make fixup` in your shell -- also run `make fix-copies` if it requests you to do so)
6. Open a PR. Tag @younesbelkada or @gante (one of us is enough)

That's it! With each change, you'll be making `transformers` a little bit better for all of us 💛 

### Models to fix:
- [x] Bart | https://github.com/huggingface/transformers/pull/21866
- [x] Bert
- [x] BigBird | https://github.com/huggingface/transformers/pull/21882
- [x] BigBirdPegasus
- [x] BioGPT | https://github.com/huggingface/transformers/pull/21844
- [x] Blenderbot
- [x] BlenderbotSmall
- [x] BlipText
- [x] Bloom
- [x] CodeGen
- [x] Esm
- [x] Git | https://github.com/huggingface/transformers/pull/21818
- [x] GPT2 | https://github.com/huggingface/transformers/pull/21772
- [x] GptNeo | https://github.com/huggingface/transformers/pull/21733
- [x] GptNeoX | https://github.com/huggingface/transformers/pull/21815
- [x] GPT-J
- [x] ImageGPT | https://github.com/huggingface/transformers/pull/21816
- [x] LED | https://github.com/huggingface/transformers/pull/21840
- [x] LongT5
- [x] M2M100 | https://github.com/huggingface/transformers/pull/21841
- [x] Marian | https://github.com/huggingface/transformers/pull/21842
- [x] MBart | https://github.com/huggingface/transformers/pull/21918
- [x] MegratronBert | https://github.com/huggingface/transformers/pull/21921
- [x] MVP | https://github.com/huggingface/transformers/pull/21920
- [x] OPT
- [x] Pegasus
- [x] PegasusX
- [x] ProphetNet | https://github.com/huggingface/transformers/pull/21772
- [x] RemBert
- [x] RoFormer
- [x] Speech2Text
- [x] Speech2Text2
- [x] SpeechT5
- [x] SwitchTransformer
- [x] T5
- [ ] TimeSeriesTransformer
- [x] TrajectoryTransformer
- [x] TrOCR
- [x] Whisper
- [x] XGLM
- [x] XLMRobertaXL
- [x] Xmod

### PR DESCRIPTION
**Pull request by pmollerus23**
# What does this PR do?

Moved gradient checkpointing clause to above the decoder layer implementation. This should fix the bug this issue addresses.

Fixes #21737 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
[GitHub Issue](https://github.com/huggingface/transformers/pull/21733)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@younesbelkada

### CODE REVIEW

File: src/transformers/models/time_series_transformer/modeling_time_series_transformer.py [hunk 1]
```diff
@@ -1118,6 +1118,13 @@ def forward(
         hidden_states = self.layernorm_embedding(hidden_states + embed_pos)
         hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
 
+        if self.gradient_checkpointing and self.training:
+                if use_cache:
+                    logger.warning_once(
+                        "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
+                    )
+                    use_cache = False
```
Comment (younesbelkada) [inline src/transformers/models/time_series_transformer/modeling_time_series_transformer.py:1126]:
```suggestion
            if use_cache:
                logger.warning_once(
                    "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
                )
                use_cache = False
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this! I left a tiny suggestion

File: src/transformers/models/time_series_transformer/modeling_time_series_transformer.py (hunk 1 above)
Comment (mollerup23) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: OK, just fixed indentation with another commit. Thanks!

Comment (mollerup23) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: Sure! For some reason, I cannot seem to install this tool properly. Are there any guides on how to install this for the Transformers repository?

Comment (younesbelkada) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: if you are inside `transformers` repo, you should be able to just run `make fixup` and it should work. What are the errors you get?

Comment (mollerup23) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: I use VSCode, and this is the output in my console when I try to run `make fixup` on the file: 

```
pmollerus@Philips-MacBook-Pro transformers % make fixup src/transformers/models/time_series_transformer/modeling_time_series_transformer.py
make: python: No such file or directory
No library .py files were modified
python utils/custom_init_isort.py
make: python: No such file or directory
make: *** [Makefile:60: extra_style_checks] Error 127
```

I used brew to install make for this project.

Comment (younesbelkada) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: I see, can you try to run `make fixup` without `src/transformers/models/time_series_transformer/modeling_time_series_transformer.py` ?

Comment (mollerup23) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: When I run that, I get this output:

```
pmollerus@Philips-MacBook-Pro time_series_transformer % make fixup
make: *** No rule to make target `fixup'.  Stop.
```

I also tried running `make fixup modeling_time_series_transformers.py` in the time series transformer directory and receive the same output.

Comment (younesbelkada) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: I see, can you set yourself on the root of `transformers` repo? i.e. where the [`Makefile` lives](https://github.com/huggingface/transformers/blob/main/Makefile), that should fix the issue I believe

Comment (mollerup23) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: Doing this, I receive the following output: 

```
pmollerus@Philips-MacBook-Pro transformers % make fixup
make: python: Command not found
No library .py files were modified
python utils/custom_init_isort.py
make: python: No such file or directory
```

Comment (younesbelkada) [reply src/transformers/models/time_series_transformer/modeling_time_series_transformer.py]: Ahh this is because python is not installed on your environment I guess, I can run the command for you if you want, just let me know

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for all your work on this!

**Review by sgugger — APPROVED** [review]
Thanks for your contribution!
